pkg/util/archive/archive.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package archive provides support for creating zip archives of local folders and returning the
+// in-memory buffer. This is how we pass Pulumi program source to the Cloud for hosted scenarios,
+// for execution in a different environment and creating the resources off of the local machine.
+package archive
+
+import (
+	"archive/zip"
+	"bytes"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/pulumi/pulumi/pkg/util/contract"
+	"github.com/pulumi/pulumi/pkg/util/logging"
+	"github.com/pulumi/pulumi/pkg/workspace"
+)
+
+// Process returns an in-memory buffer with the archived contents of the provided file path.
+func Process(path string, useDefaultExcludes bool) (*bytes.Buffer, error) {
+	buffer := &bytes.Buffer{}
+	writer := zip.NewWriter(buffer)
+
+	// We trim `path` from the pathname of every file we add to the zip, but we actaually
+	// want to ensure the files directly under `path` are not added with a path prefix,
+	// so we add an extra os.PathSeparator here to the end of the string if it doesn't
+	// already end with one.
+	if !os.IsPathSeparator(path[len(path)-1]) {
+		path = path + string(os.PathSeparator)
+	}
+
+	if err := addDirectoryToZip(writer, path, path, useDefaultExcludes, nil); err != nil {
+		return nil, err
+	}
+	if err := writer.Close(); err != nil {
+		return nil, err
+	}
+
+	logging.V(5).Infof("project archive is %v bytes", buffer.Len())
+
+	return buffer, nil
+}
+
+// Unzip a zip file into a specific directory.
+func Unzip(archive []byte, dir string) error {
```

Comment:
Seeing all of this logic makes me even more of a fan of just relying on `npm pack`.  Simple things like `.npmignore` that aren't supported here (and possibly a bunch of other little long tail things...).

Reply:
:+1: here as well, though I would be fine if we did this as an immediate follow-up to this PR.

Reply:
"It was decided" we'd try to re-use the PPC work. I was not (and am not) a fan of that strategy, and talking to Pat, the plan of record is to ship this to unblock, and then probably remove it later.